File: tests/julia/test_warnings.jl
# test_warnings.jl
# Tests for warning callback (FR-011)

include(joinpath(@__DIR__, "load_wrapper.jl"))

@testset "GIAC Wrapper Warning Tests" begin
    # The C++ class has set_warning_handler / clear_warning_handler taking a
    # std::function<void(const std::string&)>, but those methods are NOT
    # currently registered with CxxWrap in src/giac_wrapper.cpp — CxxWrap
    # does not handle the std::function argument transparently. Until the
    # wrapper exposes a compatible callback registration mechanism, these
    # tests are marked broken.

    @testset "Warning Handler Registration" begin
        ctx = GiacContext()
        @test_broken isdefined(@__MODULE__, :set_warning_handler)
    end

    @testset "Warning Handler Clear" begin
        ctx = GiacContext()
        @test_broken isdefined(@__MODULE__, :clear_warning_handler)
    end
end

File: tests/cpp/test_warnings.cpp
/**
 * @file test_warnings.cpp
 * @brief Tests for warning callback (FR-011)
 */

#include "giac_wrapper.h"
#include <iostream>
#include <cassert>
#include <string>
#include <vector>

using namespace giac_julia;

// Simple test framework macros
#define TEST(name) void test_##name()
#define RUN_TEST(name) do { \
    std::cout << "Running " #name "... "; \
    try { test_##name(); std::cout << "PASSED" << std::endl; } \
    catch (const std::exception& e) { std::cout << "FAILED: " << e.what() << std::endl; return 1; } \
} while(0)

// T063: Test warning handler registration
TEST(warning_handler_registration) {
    GiacContext ctx;
    std::vector<std::string> warnings;

    ctx.set_warning_handler([&warnings](const std::string& msg) {
        warnings.push_back(msg);
    });

    // Warning handler is set, implementation pending Phase 6
    std::cout << "(Warning handler test - pending Phase 6) ";
}

// Test warning handler clear
TEST(warning_handler_clear) {
    GiacContext ctx;

    ctx.set_warning_handler([](const std::string&) {});
    ctx.clear_warning_handler();

    // Handler should be cleared
    std::cout << "(Warning clear test - pending Phase 6) ";
}

int main() {
    std::cout << "=== GIAC Wrapper Warning Tests ===" << std::endl;

    RUN_TEST(warning_handler_registration);
    RUN_TEST(warning_handler_clear);

    std::cout << "=== All tests passed ===" << std::endl;
    return 0;
}
